metrics/graphite/graphite.go

```diff
@@ -150,16 +154,39 @@ func (e *emitter) dial() error {
 // Start will kick off a background goroutine to
 // call Flush once every interval.
 func (e *emitter) Start(interval time.Duration) error {
-	err := e.dial()
-	if err != nil {
-		return err
-	}
-	go func() {
-		for range time.Tick(interval) {
-			e.Flush()
+	var err error
+	e.start.Do(func() {
+		err = e.dial()
+		if err != nil {
+			return
 		}
-	}()
-	return nil
+		go func() {
+			t := time.Tick(interval)
+			for {
+				select {
+				case <-t:
+					e.Flush()
+				case <-e.stop:
+					return
+				}
+			}
+		}()
+	})
+	return err
+}
+
+// Stop will flush the current metrics and close the
+// current Graphite connection, if it exists.
+func (e *emitter) Stop() error {
+	if e.conn == nil {
+		return nil
+	}
+	// stop the ticking flush loop
+	e.stop <- true
+	// get one last flush in
+	e.Flush()
+	// close the connection
+	return e.conn.Close()
```

Comment:
We only know that the flush loop has received a value after line 185 has completed, we cannot assume that it completed it's call to `e.Flush()`. The local code could get there first, and could also complete `e.conn.Close()` before the flusher goroutine gets a chance to do any of the select case after receiving the value.